feat(ui): enable multiple keyboard layouts for "paste text" to remote host

dlorch opened the pull request:
Pasting text in the "Paste text" dialog sends individual key strokes to the target device using an USB keyboard emulation. Which numerical key codes are being sent depend on a mapping. This mapping was based on US English until now.

This PR provides the mechanism to enable multiple keyboard layouts for different languages and comes included with keyboard layouts for:
* Czech
* English (UK)
* English (US)
* French
* German
* Italian
* Norwegian
* Spanish
* Swedish
* Swiss French
* Swiss German

Let me know if you would like to see further keyboard layouts and I see what I can do.

I've seen #116, but the scope of that PR is much bigger and aiming at overhauling the entire keyboard functionality. For this PR here, I wrote the minimal set of changes to enable keyboard layouts in the "Paste text" dialog. We can then later align the initiatives.

This provides fixes for #22, #30, #65, #104, #47.

Could potentially address #436.

![paste-modal-spanish](https://github.com/user-attachments/assets/96b2a9ac-80c1-4d8d-93de-d69e453b0a40)

![keyboard-layout-config](https://github.com/user-attachments/assets/714be44f-a8f4-45d9-99c7-0b017d6e9854)

IDisposable commented on ui/src/keyboardLayouts.ts:
```diff
@@ -0,0 +1,42 @@
+export type KeyCombo = KeyInfo & { deadKey?: boolean, accentKey?: KeyInfo }
+
+export const layouts = {
+  "cs_CZ": "Czech",
```
Should these really be in the language they represent? For example `"de_DE": "Deutsch",

dlorch replied:
Yes it could be done and would look like something like this for the above:

```javascript
import { chars as chars_cs_CZ, code as code_cs_CZ, name as name_cs_CZ } from "@/keyboardLayouts/cs_CZ"

export const layouts = {
  code_cs_CZ: name_cs_CZ,
   ..
}
```

It wouldn't save any characters, but it would make it cleaner in terms of moving the definitions together into the keyboard layout files. Is that what you had in mind?

IDisposable reviewed (approved):
Generally looks awesome!

dlorch commented:
> Generally looks awesome!

Thank you for the feedback and for reviewing! My head is still spinning from all the different amazing accent variants that are available in the Czech language.

dlorch commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -42,22 +61,39 @@ export default function PasteModal() {
     try {
       for (const char of text) {
-        const { key, shift } = chars[char] ?? {};
+        if (!keyboardLayout) continue;
```
`keyboardLayout` is filled in async via `getKeyboardLayout`. In a (very unlikely, but still probable) case, the state of that variable could be populated and change at the same time while the looping over the character occurs. But it would make for an odd behavior (pasting would only work for a portion of the text, not all).

I agree it makes sense to move the guard statements outside of the loop.

IDisposable commented:
This is almost perfect, the only remaining thing (in my opinion) is to make the name of the locale be IN the language of the locale. I've submitted a PR on your repo @dlorch https://github.com/dlorch/kvm/pull/1

dlorch commented:
> This is almost perfect, the only remaining thing (in my opinion) is to make the name of the locale be IN the language of the locale. I've submitted a PR on your repo @dlorch [dlorch#1](https://github.com/dlorch/kvm/pull/1)

Merged!

IDisposable commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -37,27 +56,43 @@ export default function PasteModal() {
-            },
-          );
-        });
+                send("keyboardReport", hidKeyboardPayload([], 0), params => {
```
Just trying to ensure I get this... the purpose here is to send no actual scan-codes, but send a 0 for the modifier to un-shift/un-Ctrl/un-Alt... correct?

dlorch replied:
Modifiers aren't sticky. The toggle state of the modifiers need to be sent along each invocation of the Linux USB HID gadget driver. Here's how the modifier mask is built up in the [hid_gadget_test.c](https://www.kernel.org/doc/Documentation/usb/gadget_hid.txt) example program:

```c
		for (i = 0; kmod[i].opt != NULL; i++)
			if (strcmp(tok, kmod[i].opt) == 0) {
				report[0] = report[0] | kmod[i].val;
				break;
			}
```

From what I can tell, the highlighted code will result in an array of zeros being sent to the Linux USB HID gadget as filled here:

https://github.com/jetkvm/kvm/blob/94f36d0facfac6f15e980c934d2c7c0c2cc1d92c/internal/usbgadget/hid_keyboard.go#L84-L88

The net result should be a "no-op".

This line appeared in my change set due to re-indentation, it is a snippet from the original code. Great to point this one out. I don't know what it does. It should be harmless, but could potentially be removed. I will run the code and have a closer look.

Maybe the original author (@adamshiervani) remembers more about the purpose of this code fragment.

dlorch replied:
I've run it. I entered `c` in to the paste text box, and had `ccccccccccccccccccccccccccc` coming out on the target device, until I pressed a button on my keyboard to end it.

Forget my above comments, this is the equivalent of a "key up" event.

We could argue about moving this outside of the loop, but it's probably better the way it is (there's a "key up" event after each sent character, it avoids duplication of key presses).

dlorch replied:
I found it! It's the equivalent of this behavior for "hold" in [hid_gadget_test.c](https://www.kernel.org/doc/Documentation/usb/gadget_hid.txt):

```c
			if (!hold) {
				memset(report, 0x0, sizeof(report));
				if (write(fd, report, to_send) != to_send) {
					perror(filename);
					return 6;
				}
			}
```

An argument could be made to move handling of the "hold" behavior into the "backend" https://github.com/jetkvm/kvm/blob/94f36d0facfac6f15e980c934d2c7c0c2cc1d92c/internal/usbgadget/hid_keyboard.go

Thoughts?

adamshiervani replied:
> An argument could be made to move handling of the "hold" behavior into the "backend" [hid_keyboard.go](https://github.com/jetkvm/kvm/blob/94f36d0facfac6f15e980c934d2c7c0c2cc1d92c/internal/usbgadget/hid_keyboard.go).

Agree!

> This would require changing the signature of the [RPC handler](https://github.com/jetkvm/kvm/blob/63c2272c4589c9b052ee75d5496a555af1a1632b/jsonrpc.go#L994) to:

If we go down this path, we need to be mindful when changing the interface. For local access, this won’t be an issue since the new JS is embedded in the Go binary and will always match. The challenge arises when the Cloud JS is a newer version than the device app. It’s nothing we can’t work around -- just something to keep in mind.

dlorch commented:
@ym I received a notification of a `lint` error in `usb_mass_storage.go`: https://github.com/jetkvm/kvm/actions/runs/15002349526 but this file is not part of my changeset.

Could you try again? I have removed the `git merge dev` from the history and force pushed the changes to this branch.

Since you seem to be squashing the feature branches when merging, I think that `git merge dev` could have caused some issues.

Am I expected to resolve merge conflicts to `dev`? If yes, through rebasing with `git rebase dev`?

dlorch commented:
@IDisposable do you happen to have advice on how to deal with merge conflicts with `dev`? Since this PR was created, the `dev` branch has evolved a bit. The merge conflicts are not structural, but just having a change in the same code area than another contributor can cause them.

What is the maintainers' preferred approach:

1. do I need to `git merge dev` and resolve the conflicts? I think this is unlikely, as PRs are squashed when merged
2. do I need to run `git rebase dev` and resolve the conflicts?
3. do you fix the (simple) merge conflicts yourself?

Many thanks in advance!

IDisposable commented:
I'll pull yours and resolve the issues and push back as a PR to your repo... probably not until Friday

dlorch commented:
@IDisposable I rebased from dev and fixed the merge conflicts coming from #449.

IDisposable commented on jsonrpc.go:
```diff
@@ -991,7 +1003,7 @@ var rpcHandlers = map[string]RPCHandler{
 	"setNetworkSettings":     {Func: rpcSetNetworkSettings, Params: []string{"settings"}},
 	"renewDHCPLease":         {Func: rpcRenewDHCPLease},
-	"keyboardReport":         {Func: rpcKeyboardReport, Params: []string{"modifier", "keys"}},
+	"keyboardReport":         {Func: rpcKeyboardReport, Params: []string{"modifier", "keys", "hold"}},
```
I wonder if we need to worry about backward compatibility and use a second endpoint? e.g. leave the original line there and add another `"keyboardReport2": ` for the new api?  @ym?

dlorch replied:
This caused issues with "Shift" key handling. I have reverted.

IDisposable commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -8,14 +8,21 @@ import { GridCard } from "@components/Card";
+  return { keys, modifier, hold };
 };
 
+const modifierCode = (shift?: boolean, altRight?: boolean) => {
```
I know this is meant to address the AltGr key specifically, but I wonder at the keyboard level if we shouldn't just be sending ALL the modifiers through to allow any keystroke make it through. Something like mapping and ORing together ALL of the bits (for general purpose use, not just in _PasteModal.tsx_):

```typescript
const modifierCode = (
   leftShift?: boolean, rightShift?: boolean, 
   leftControl?: boolean, rightControl? boolean, 
   leftAlt?: boolean, rightAlt?: boolean,
   leftMeta?: boolean, rightMeta?: boolean
) => {
  return (
       leftShift ? modifiers["ShiftLeft"] : 0 | rightShift ? modifiers["ShiftRight"] : 0
       | (leftAlt ? modifiers["AltLeft"] : 0 | (rightAlt ? modifiers["AltRight"] : 0
       | (leftCntrl ? modifiers["ControlLeft"] : 0 | (rightCntrl ? modifiers["ControlRight"] : 0
       | (leftMeta ? modifiers["MetaLeft"] : 0 | (rightMeta ? modifiers["MetaRight"] : 0
 )
}
```
_warning, GitHub editor syntax check... this is just a concept, not an implementation_

dlorch replied:
It could be done, but none of the keyboard layouts as part of this PR seem to requite Ctrl or Alt. The original code only supported Shift.

What keys would require Ctrl or Alt?

IDisposable replied:
It's more about the general support (where as this handles the pasting logic)... I might be conflating for no reason... but I got bit just yesterday by Shift-Backspace (which is something you would never "paste") not working... which for MacOS is "delete right"... and so I've been thinking about how we can become agnostic to the _application_ at the other end of the virtual keyboard.

Again, this comment really doesn't apply specifically to the Paste function, sorry

dlorch replied:
Mark, as it stands, do you see any blockers or concerns to get this PR merged?

We’ve moved the layout name variables into the layout files together, I’ve swiftly fixed the „hold key“ issue that you raised and also got all merge conflicts out of the way.

Getting this PR merged would allow people to try it out and potentially report bugs on it, and also give other users the opportunity to contribute new keyboard layouts.

Refactoring is still possible afterwards and should be quick and easy.

IDisposable replied:
I think it's good to go from a "stare at code" standpoint, but I have not yet built and run on my Jet ... will do that this weekend

IDisposable commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -37,27 +56,40 @@ export default function PasteModal() {
     if (rpcDataChannel?.readyState !== "open" || !TextAreaRef.current) return;
+    if (!keyboardLayout) return;
+    if (!chars[keyboardLayout]) return;
 
```
Let's assign the lookup and use that lower down like:
```typescript
    if (!keyboardLayout) return;
   const layoutChar = chars[keyboardLayout];
    if (!layoutChars) return;
```

dlorch replied:
Sure, could you send a PR?

IDisposable commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -37,27 +56,40 @@ export default function PasteModal() {
     try {
       for (const char of text) {
-        const { key, shift } = chars[char] ?? {};
+        const { key, shift, altRight, deadKey, accentKey } = chars[keyboardLayout][char]
```
... then here and elsewhere later down do 
```typescript
const { key, shift, altRight, deadKey, accentKey } = layoutChars[char]
```

IDisposable commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -113,7 +145,7 @@ export default function PasteModal() {
                             [...new Intl.Segmenter().segment(value)]
                               .map(x => x.segment)
-                              .filter(char => !chars[char]),
+                              .filter(char => !chars[keyboardLayout][char]),
```
```typescript
.filter(char => !layoutChars[char]),
```

IDisposable commented on ui/src/keyboardLayouts/en_UK.ts:
```diff
@@ -0,0 +1,107 @@
+import { KeyCombo } from "../keyboardLayouts"
```
Could this (or the US) be synthesized from the other one (just overlaying as is done for _Français de Suisse_ deriving from _Français_? Could also do so for German and Swiss German.

dlorch replied:
Sure, we could build an inheritance tree to see which keyboard layouts are variants of the others. But it seems a lot of effort to maintain and would require a lot of brain muscle to remember the dependencies when future changes are being made to not introduce regressions, I think.

But I‘m open to it, feel free to send a PR.

IDisposable commented on ui/src/keyboardMappings.ts:
```diff
@@ -43,7 +43,7 @@ export const keys = {
   Home: 0x4a,
   Insert: 0x49,
-  IntlBackslash: 0x31,
+  IntlBackslash: 0x64,
```
I was going to correct that when I was doing the cleanup/refactor and forgot. Thanks!

dlorch replied:
Thanks for the review!

dlorch commented:
> Is there an option to add Dutch (Belgium) to the keyboards?

That would be the AZERTY keyboard as used in Flanders? https://kbdlayout.info/kbdbene

dlorch commented:
> > > Is there an option to add Dutch (Belgium) to the keyboards?
> > 
> > 
> > That would be the AZERTY keyboard as used in Flanders? https://kbdlayout.info/kbdbene
> 
> Correct! 👍

It's added https://github.com/jetkvm/kvm/pull/405/commits/96985645506ab94d6d996cea5272515bf9ab955b

dlorch commented:
@IDisposable I rebased from dev again and fixed the merge conflicts. I would consider this PR ready. Let's ship this thing! The many users in https://github.com/jetkvm/kvm/issues/22, https://github.com/jetkvm/kvm/issues/30, https://github.com/jetkvm/kvm/issues/65, https://github.com/jetkvm/kvm/issues/104, https://github.com/jetkvm/kvm/issues/47 will be thankful. Thank you for the support!

dlorch commented:
@IDisposable could you rebase from my branch and not merge? I‘m seeing merges in your feature branch and that‘s not going to work: https://github.com/IDisposable/kvm/commits/paste-text-keyboard/

I‘ve rebased from the „dev“ branch to keep current and fix merge conflicts. Rebasing rewrites history and you‘re basing off an old branch.

My recommendation:
1. Get my PR merged in as-is, as it will apply cleanly (but „dev“ is moving fast with changes coming in, so please don‘t hold off. Waiting means more work for me…)
2. Create separate PRs for your changes based off the new dev branch

FYI https://www.atlassian.com/git/tutorials/merging-vs-rebasing

dlorch commented:
@IDisposable for illustration, this is the commit graph of my feature branch:

<img width="829" alt="Screenshot 2025-05-20 at 11 38 39" src="https://github.com/user-attachments/assets/d0440d76-fe79-44e7-b76c-0d06a8c3ca8a" />

What can we see on the picture? My feature branch "branches off" the HEAD of the "dev" branch (b4dd496) in a straight line.

I was able to do this by "rebasing" my changes on top of the latest "dev" branch. This is a technique to rewrite the git history as if I would have created a fresh branch just now.

All operations that rewrite history also change the commit IDs (hashes), and therefore there needs to be a communication and alignment between developers that work on the same branch, otherwise people will continue to work off different histories of the same code. Maybe the message got lost in translation, but I've inquired above [on your preferred approach](https://github.com/jetkvm/kvm/pull/405#issuecomment-2882803086) to get the merge conflicts resolved. Apologies if I was not clear on the implications of this choice.

The JetKVM maintainers seem to "squash" feature branches before merging them. "Squashing" is a technique to "flatten" a list of commits into a single one. This helps to keep the Git history compact and is generally a recommended approach, because after merging a PR, the detailed feature branch can be thrown away and only the squashed commit will be kept. But this also means the feature branches should be "clean" and ideally not contain any unrelated changes that would occur from for example merging things.

For comparison, this is the commit graph of your feature branch:

<img width="807" alt="Screenshot 2025-05-20 at 11 36 29" src="https://github.com/user-attachments/assets/e2d58692-2736-4ae7-9a2e-0055c1c77fac" />

Your feature branch now contains a mix of my changes, your changes, and a few unrelated commits from the latest "dev" branch. That could cause complications when "squashing" the branch.

Also pay close attention to the "SHA" column between our two commit graphs. For the same commits, we have different values, as I have rebased for the second time and you are working off an older (outdated) branch.

Mark, you are a proponent of keeping PRs "low risk", small and self-contained. For all intents an purposes, this PR is ready to be merged for already two weeks. As mentioned before, changes that occur in "dev" that affect the same code areas as this PR (happened two times already) mean there's effort for me to do an OSS contributor to keep this PR up to date.

Unless there are reasons not to do so, let's get this PR merged and fix forward. Any refactorings that are based on top of it, can be cleanly applied as separate and "low risk" PRs. I hope this is a reasonable request.

erialor commented:
Would dk-latin1 keyboard be possible, please :)

dlorch commented:
> Would dk-latin1 keyboard be possible, please :)

Just to reconfirm, it would be this Danish keyboard layout? https://www.kbdlayout.info/kbdda

erialor commented:
> > Would dk-latin1 keyboard be possible, please :)
> 
> Just to reconfirm, it would be this Danish keyboard layout? https://www.kbdlayout.info/kbdda

Excatly :)

dlorch commented:
@IDisposable again I hope you and your loved ones are doing well.

If I had to summarize my efforts on this PR, it would be about 30% to write the code, which I‘d say is quite a feat. considering the number of keyboard layouts already included (and manually validated), supporting complex accent key handling, and 70% going back and forth on variable naming, shuffling logic around and git branching woes.

The PR proposed by you https://github.com/dlorch/kvm/pull/2 still contains unrelated changes merged in from „dev“, duplicates logic in areas, contains debug statements, and there‘s also shift in the programming style versus what‘s already there.

This PR here is ready to be merged, has clear user value, but instead it is being micro-managed without clear outline on how to get this into „dev“. I‘ve been extremely responsive on getting everything addressed that was raised to me, but it has become a never ending infinite loop of iterations on refactoring stuff.

I‘m doing these OSS contributions in my free time because I like the product and need this feature here to be fixed for myself, I’ve been also nice enough to cover other languages, but I can‘t support the sustained back and forth on this anymore.

At this point, this is a take-it-or-leave-it situation. I hope this reasonable and understandable.

Looking forward to the Google Meet and having a chat!

CC: @adamshiervani @ym

IDisposable commented:
I'm fine with whatever the decision is... I can always submit a new one on top of this after it gets merged. I understand you concerns about a different coding style (in that I'm using a distinct type for the keyboard layouts so we get all the attributes in one object instead of having to keep two things in sync). The changes from `dev` are because I rebased onto _your_ version, which was recently rebased as well, not something else I pulled in (unless unintentionally).

adamshiervani commented on ui/src/keyboardLayouts/sv_SE.ts:
```diff
@@ -0,0 +1,164 @@
+  V: { key: "KeyV", shift: true },
+  W: { key: "KeyW", shift: true },
+  X: { key: "KeyX", shift: true },
+  Y: { key: "KeyZ", shift: true },
+  Z: { key: "KeyY", shift: true },
```
Flip

adamshiervani commented on ui/src/keyboardLayouts/sv_SE.ts:
```diff
@@ -0,0 +1,164 @@
+  v: { key: "KeyV" },
+  w: { key: "KeyW" },
+  x: { key: "KeyX" },
+  y: { key: "KeyZ" },
+  z: { key: "KeyY" },
```
Flip

adamshiervani commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -27,6 +34,18 @@ export default function PasteModal() {
+      if ("error" in resp) return;
+      setKeyboardLayout(resp.result as string);
+    });
+  }, []);
```
Don't forget the dependencies

adamshiervani commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -37,27 +56,43 @@ export default function PasteModal() {
     setPasteMode(false);
     setDisableVideoFocusTrap(false);
     if (rpcDataChannel?.readyState !== "open" || !TextAreaRef.current) return;
+    if (!keyboardLayout) return;
```
Don't forget to add to the useCallback dependencies

adamshiervani commented on ui/src/components/popovers/PasteModal.tsx:
```diff
@@ -132,6 +167,11 @@ export default function PasteModal() {
                 </div>
+		<div className="space-y-4">
+                  <p className="text-xs text-slate-600 dark:text-slate-400">
+                    Sending key codes using keyboard layout {layouts[keyboardLayout]}
```
Nit: 
- "Sending text using keyboard layout: XXXX"
- Could you embolden the language

adamshiervani reviewed (changes requested):
Wonderful work! I tested it on my dev setup, and it works great functionally. Just a couple of tiny change requests.

Let's get this one shipped soon. I know many people will be happy!

IDisposable commented on ui/src/keyboardLayouts/sv_SE.ts:
```diff
@@ -0,0 +1,164 @@
+  "$": { key: "Digit4", altRight: true },
+  5: { key: "Digit5" },
+  "%": { key: "Digit5", shift: true },
+  6: { key: "Digit6" },
```
Missing `€` which is the altRight of the `5` key

https://kbdlayout.info/KBDSW/

dlorch replied:
According to the layout, it should be in two places: **Alt Gr + 5** or **Alt Gr + e**

I did try this out and only one of them worked on my system.

It's there already in the file:

```
  "€": { key: "KeyE", altRight: true },
```

IDisposable replied:
Totally missed the AltGr+e version, sorry.

IDisposable commented on ui/src/keyboardLayouts/sv_SE.ts:
```diff
@@ -0,0 +1,164 @@
+  j: { key: "KeyJ" },
+  k: { key: "KeyK" },
+  l: { key: "KeyL" },
+  m: { key: "KeyM" },
```
Missing `µ` which is the altRight of the `m` key https://kbdlayout.info/KBDSW/

dlorch commented:
> > > Would dk-latin1 keyboard be possible, please :)
> > 
> > 
> > Just to reconfirm, it would be this Danish keyboard layout? https://www.kbdlayout.info/kbdda
> 
> Excatly :)

@erialor please subscribe to notifications on this PR: https://github.com/jetkvm/kvm/pull/494

IDisposable commented:
YAY! Nice to see this merged :)

dlorch commented:
> YAY! Nice to see this merged :)

Yeah we did it! Nice one, thanks for the support.

v7o73 commented:
Hallo Daniel, das funktioniert super mit deiner Erweiterung, das rettet alles :-)
Ein grosses Merci !

IDisposable commented:
Yep, those two things were built in parallel so we need to implement these fixes for Paste (and Virtual Keyboard needs adjusting once #500 is merged).

We need to further unify the keyboard state management so things can say "I want to emit this keycode" (for macros and virtual keyboard) or "I want to cause this unicode to be entered" (for paste and macros in literal mode).
